src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs

```diff
@@ -0,0 +1,249 @@
+
+            var completionList = await completionService.GetCompletionsAsync(
+                document,
+                triggerLocation,
+                GetRoslynTrigger(trigger, triggerLocation)).ConfigureAwait(false);
+
+            if (completionList == null)
+            {
+                return new EditorCompletion.CompletionContext(ImmutableArray<EditorCompletion.CompletionItem>.Empty);
+            }
+
+            var filterCache = new Dictionary<string, EditorCompletion.CompletionFilter>();
+
+            var itemsBuilder = ImmutableArray.CreateBuilder<EditorCompletion.CompletionItem>();
+            foreach (var roslynItem in completionList.Items)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var item = Convert(document, roslynItem, completionService, filterCache);
+                item.Properties.AddProperty(TriggerBuffer, triggerLocation.Snapshot.TextBuffer);
+                if (!string.IsNullOrEmpty(item.DisplayText))
+                {
+                    itemsBuilder.Add(item);
+                }
+            }
+
+            var items = itemsBuilder.ToImmutable();
+
+            var suggestionItemOptions = completionList.SuggestionModeItem != null
+                    ? new EditorCompletion.SuggestionItemOptions(
+                        completionList.SuggestionModeItem.DisplayText,
+                        (completionList.SuggestionModeItem.Properties.TryGetValue(Description, out var description)
+                            ? description
+                            : string.Empty))
+                    : null;
+
+            session.Properties.AddProperty(MustSetSelection, suggestionItemOptions != null);
+
+            return new EditorCompletion.CompletionContext(
+                items,
+                suggestionItemOptions,
+                suggestionItemOptions == null ? EditorCompletion.InitialSelectionHint.RegularSelection : EditorCompletion.InitialSelectionHint.SoftSelection);
+        }
+
+        public async Task<object> GetDescriptionAsync(IAsyncCompletionSession session, EditorCompletion.CompletionItem item, CancellationToken cancellationToken)
+        {
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(RoslynItem, out var roslynItem) ||
+                !item.Properties.TryGetProperty<ITextBuffer>(TriggerBuffer, out var triggerBuffer))
+            {
+                return null;
+            }
+
+            var document = triggerBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return null;
+            }
+
+            var service = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
+            var description = await service.GetProvider(roslynItem).GetDescriptionAsync(document, roslynItem, cancellationToken).ConfigureAwait(false);
```

Comment:
this is *Extremely* unsafe.  you are asking the item to answer a questoin based on the current state of the doc, not the original state when the item was created.  Taht item may have stored a position in it that is no longer valid at this time.  If you want to do this, store the ITextSnapshot and make sure you can go back from that to Document.

Reply:
> We take the position from the snapshot obtained from TriggerBuffer

Where?  The code does this:

```var document = triggerBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();```

that's using CurrentSnapshot, not hte trigger snapshot.

Reply:
Thank you! It seems we should store the snapshot not the buffer.

---
In reply to: [222481165](https://github.com/dotnet/roslyn/pull/29016#discussion_r222481165) [](ancestors = 222481165)